Sass/spin sampling beccatini palermo formula

nilssass opened the pull request:
This PR enables the sampler to sample spin 4-vectors for all particles given the vorticity file from vHLLE. The spin sampling can be de/activated with the corresponding config key. To do so I have introduced the following new config keys:

- `sample_spin` _(boolean)_: Activates or deactivates spin sampling
- `vorticity_file` _(string)_: Path to the vorticity file from vHLLE. If not given explicitely, the sampler expects it to be located in the same directory as the freezeout surface
- `create_vorticity_vector_output`_(boolean)_: Creates an additional output file storing all vorticity vectors of all particles

Additionally, this PR includes unit tests for all new files and functions.

Carl-Rosenkvist commented on src/include/vorticity.h:
```diff
@@ -0,0 +1,66 @@
+  // Default constructor nulls all vorticity components
+  Vorticity() : vorticity_({0.0}) {}
+
+  using reference = double&;
```
Why use alias for double?

Carl-Rosenkvist commented on src/spin.cpp:
```diff
@@ -0,0 +1,193 @@
+        denominator += denominator_term;
+      }
+
+      if (std::abs(denominator) < 1e-8) {
```
Check if suffecnetly negative but you only use the absolute value

nilssass commented on src/vorticity.cpp:
```diff
@@ -0,0 +1,307 @@
+    // The constructor already sets all components to 0.
+    surf[i].vorticity = std::make_unique<Vorticity>();
+
+    // TODO: Set energy density in corona cells
```
What about the TODO?

Carl-Rosenkvist commented on src/include/vorticity.h:
```diff
@@ -0,0 +1,66 @@
+  const_reference at(int i, int j) const { return vorticity_[i * 4 + j]; }
+
+  // Boost the vorticity tensor to the fluid rest frame with given boost matrix
+  void boost_vorticity_to_fluid_rest_frame(const double (&boostMatrix)[4][4]);
```
Why is this a c-style array? can std::array be used here?

Carl-Rosenkvist commented on src/spin.cpp:
```diff
@@ -0,0 +1,193 @@
+        "Theta cannot be calculated for massless particles.");
+  }
+  // Transform the row and column indices for a 4x4 matrix into a 1d index
+  auto at = [](int i, int j) { return (i * 4 + j); };
```
should this be a lambda? and is it tested?

Carl-Rosenkvist commented on tests/spin.cpp:
```diff
@@ -0,0 +1,467 @@
+
+using namespace spin;
+
+// Precomputed lookup table for Levi-Civita symbol
```
should not this be in const.h? or is this just for the test?

Carl-Rosenkvist reviewed (approved):
In general the code is well written and well documented. There is also good tests. I have some minor comments about changes that I personally would like but are not necesarry. Hence, I approve this pull request.

yukarpenko commented:
Thanks for the update. I'll need to check the formula first. What makes it a bit difficult is that, to my understanding, this branch contains a lot of technical/cosmetic changes to the code with no change in the numerics. Why didn't you somehow differentiate it and push the technical/cosmetic changes first, and then the addition of spin sampling would be a separate addition. Last time we had a chat, my understanding was that Robin was, generally speaking, polishing the sampler code. I am asking this since I only interact with you guys here, and I have no idea what kind of interactions you have in-person. So it would be great if I understand what is going on in general.

yukarpenko commented:
My apologies again - I still didn't reach this change before leaving for holidays. So I'll try to check the earliest I can, but it can be anywhere from tomorrow until 3 weeks later. Parenthood introduces significant fluctuations to work time. Sorry.

RobinSattler commented:
@nilssass, it would be great if you could add a brief description of the new keys to the README file that we have at least a minimal overview of the features of the sampler in place. :nerd_face:

yukarpenko commented:
ah, it was me! Well, likewise, my old comment is wrong. Thanks for digging into the history with git blame - I did not do beyond the blame button on github, since I do not really need to spend time on this.
The connection is that I am looking at the code. the PR shows only diffs which are not enough sometimes.

yukarpenko commented:
So @nilssass a question about formulas, as I am progressing through the essential parts of the code. The Lorentz boost matrix, which is boostMatrix[][], is supposed to boost vectors/tensors with upper indices. That is why, for example, hydro prints d\sigma_mu (a vector with lower index), so in the sampler, as  I load d\sigma_mu from the freezeout file, it is initially a vector with lower index (note _mu). Then I convert it to a vector with upper index:
```
    dsigma.SetXYZT(-surf[n].dsigma[1], -surf[n].dsigma[2], -surf[n].dsigma[3],
                   surf[n].dsigma[0]);
```
 then I boost it to the LRF
```
     dsigma.Boost(-surf[n].u[1] / surf[n].u[0], -surf[n].u[2] / surf[n].u[0],
                 -surf[n].u[3] / surf[n].u[0]);
```
then I convert it again to a vector with lower index:
```
    surf[n].dsigma[0] = dsigma.T();
    surf[n].dsigma[1] = -dsigma.X();
    surf[n].dsigma[2] = -dsigma.Y();
    surf[n].dsigma[3] = -dsigma.Z();
```
later on, you can see how the pi^{mu nu} is boosted to the LRF:
```
...
              _pi[index44(i, j)] += surf[n].pi[index44(k, l)] *
                                    boostMatrix[i][k] * boostMatrix[j][l];
...
```
which is correct since _pi[index44(i, j)] represents a rank-2 tensor with upper indices.
Now, I am not sure whether it is like that with the vorticity. in fld.cpp in vHLLE, the function Fluid::printDbetaHeader() tells me that hydro writes d_mu beta_nu, which is a tensor with two lower indices. Then, in the sampler you read this in a function set_vorticity_in_surface_cells(), then there is a function Vorticity::boost_vorticity_to_fluid_rest_frame() which seems to boost the vorticity as if it is a tensor with upper indices. Am I missing something or is it a mistake?

yukarpenko commented:
Next thing:  (I will be posting partial comments/questions, since I do not have a continuous time slot to have a look at the entire code)
In `Vorticity::boost_vorticity_to_fluid_rest_frame` , and potentially in other places in the code, you are creating 1D 16-wide arrays to represent a 4x4 matrix (vorticity tensor), but then to get particular components of the vorticity tensor, you are accessing the elements directly as `vorticity_copy[k * 4 + l]` . If you are sure you have control what the first index and what is second index there, it is good (then, no mistakes), but otherwise to my taste it is prone to errors. That's why I am using a helper function `index44(k, l)` always, when accessing such 1D arrays, to be sure that the first and the second indices in the underlying 4x4 matrix are always in the same order throughout the code. In principle, would be good if you use the same function, otherwise I hope you have the indices under control.

yukarpenko commented:
Next things I spotted, in calculate_and_set_spin_vector():
- const double mu = freezeout_element.mub;
  Here you are assuming the chemical potential is only baryon one, so you're omitting muQ and muS. I guess it would be (mostly) OK if you are only considering baryons, but why not to write an expression for the "full" chemical potential?
- const double distribution_argument = (energy_density - mu) / temperature;
  Emm... energy_density and mu should have different units, and I think what should be there is not the energy density but the energy of the particle. It looks like a basic mistake in the formula, but maybe I am confused about the notations?
- I am getting a bit confused about the upper/lower index situation, since I found a comment
 "...As we only
  // get quantities with upper indices from the sampler, we need to lower these..."
  Which caught me by surprise since, as a physicist, I would read the header line 
            "T  μB  μQ  μS  ∂₀β₀  ∂₀β₁  ∂₀β₂  ∂₀β₃  ∂₁β₀  ∂₁β₁  ∂₁β₂  ∂₁β₃  "
          "∂₂β₀  ∂₂β₁  ∂₂β₂  ∂₂β₃  ∂₃β₀  ∂₃β₁  ∂₃β₂  ∂₃β₃  ϵ";
 as the one containing thermal vorticity with both lower indices.
I'll need to check it on the side of vHLLE later (end of work today). If it is really the case that vHLLE writes d^mu beta^nu, then my earlier comment about the Lorentz boost is not relevant.

yukarpenko commented:
So, vHLLE outputs the thermal vorticity, \varpi_{\mu\nu}, with both lower indices. I am guessing my first formula comment is correct after all. That's all I spotted in the formulas for now. Maybe we discuss all of that somehow?
Last thing: I didn't really think about it much, so I may be wrong here, but still. What you are computing for each hadron - Eq. 60 in the references you're using (or any other similar formula in past works) is actually a mean (average) polarization, which shouldn't really be a property of a particle but a property of an ensemble of particles. Wouldn't you need to actually sample the spin orientation based on the averages you've computed?

yukarpenko commented:
A related but not directly connected thing: when I started to check the formulas that are coded in, I realised that it might be a good time to put more comments in the existing part of the sampler, to clearly note which quantities are stored in the variables (the lower-upper index story, for example). But then once I started to add the comments, I realised that it may make the merging harder, so once we sort out with the formulas you'vr coded in, and merge, I would add some explanatory comments finally (if I don't forget by then).

nilssass commented:
Dear @yukarpenko thank you for your detailed review. Indeed you spotted two issues with the code. I will quickly give you an overview over the changes/fixes I have done:

## Incorrect index positions
In `gen.cpp` [line 215](https://github.com/smash-transport/smash-hadron-sampler/blob/699c047e181f94342acd47e07e4bc18896fb167b/src/gen.cpp#L215) you can see that I have added the line 
```c
boostMatrixCov = create_covariant_boost_matrix(boostMatrix);
``` 
which switches the positions of the boost matrix' indices such that the new boost matrix acts on lower indices. This "covariant" boost matrix is now used to boost vorticity. The definition of the function can be found in `gen.cpp`, [line 77 ff](https://github.com/smash-transport/smash-hadron-sampler/blob/699c047e181f94342acd47e07e4bc18896fb167b/src/gen.cpp#L77). 

**Test:**
I have also added a test in `tests/gen.cpp` to ensure the correct behaviour of the new function and the resulting boost matrix. [See here](https://github.com/smash-transport/smash-hadron-sampler/blob/699c047e181f94342acd47e07e4bc18896fb167b/tests/gen.cpp#L82)

## Mismatching units
You are right in pointing out that the units in the distribution argument in `spin.cpp`, [line 139](https://github.com/smash-transport/smash-hadron-sampler/blob/699c047e181f94342acd47e07e4bc18896fb167b/src/spin.cpp#L139):
```c
const double distribution_argument = (energy_density - mu) / temperature;
```
are not matching. Indeed this should be the energy of the sampled particle. I have replaced the energy density in this and all other similar lines.

## Using the full chemical potential
In `spin.cpp`, [line 139](https://github.com/smash-transport/smash-hadron-sampler/blob/699c047e181f94342acd47e07e4bc18896fb167b/src/spin.cpp#L110) I am using a new function `chemical_potential` which calculates the "full" chemical potential per particle, following $\mu_i = B_i \mu_B + Q_i \mu_Q + S_i \mu_S$

**Test:**
The function is tested in `tests/spin.cpp`, [line 487 ff](https://github.com/smash-transport/smash-hadron-sampler/blob/699c047e181f94342acd47e07e4bc18896fb167b/tests/spin.cpp#L487) to guarantee it's correct behaviour.

## Cosmetics
You were raising concerns about the explicit access of matrix elements of the function `boost_vorticity_to_fluid_rest_frame` in `vorticity.cpp`. Even though there is an explicit test that ensures the correct behaviour of the matrix multiplication, I have changed it to a lambda in [line 293](https://github.com/smash-transport/smash-hadron-sampler/blob/9ba00eec020a15edafd8902e11060af5280a25d0/src/vorticity.cpp#L293)

---

With these changes we should be ready to merge.

nilssass commented:
@yukarpenko Good question. The equations are frame independent , yes. I’m doing it in the fluid rest frame mainly to stay consistent with how the code has always worked and to avoid surprising users. `gen::load` boosts vorticity to the fluid rest frame the same way as it does for dsigma/pi^{mu, nu}..., and the spin 4-vector is finally boosted to the lab frame exactly like the 4-momentum. The only new boost I added lives inside `spin::calculate_and_set_spin_vector(...)` just to make sure all inputs are in the same frame, so just a technical alignment.
Probably we could do everything in a global frame, but that would mean touching a bunch of places and re-threading the data flow, with a real risk of introducing new bugs. In the current setup, it mirrors the old logic, needs the least surgery, and I’ve tested that it gives the correct results. Therefore, I would really avoid refactoring it now, since the current PR is well tested now.

yukarpenko commented:
BTW I don't see how momentum sampling in FRF and the rest of machinery in global frame would be confusing. The spin calculation seems to be logically separate from the momentum generation, done in a different module.
